### PR TITLE
fix(template-conditions): replace backtracking string tokenization with a linear scanner

### PR DESCRIPTION
**Pull request by jan-kubica**
Fixes the CodeQL `js/polynomial-redos` finding in `packages/template-conditions/src/parse.ts`.

## Root cause

The tokenizer matched quoted strings via an unanchored `matchAll` over the whole expression. On an unterminated string literal, the failed scan for a closing quote is retried starting at every subsequent `"` position, which is quadratic on adversarial input (e.g. `"` followed by repeated `\"\a`). Notably, the textbook rewrite of the nested-quantifier string pattern is not sufficient: with `matchAll`'s retry behavior it still measured ~1 s at 16 KB of input.

## Fix

- String literals are tokenized by a dedicated single-pass scanner (`scanString`) that visits each character at most once, entered explicitly when the tokenizer sees `"`. No backtracking or positional retry is possible by construction.
- Remaining token alternatives stay in one regex, now matched with the sticky flag at an explicit position; none of them contain an unbounded quantifier that depends on a possibly-absent terminator.
- One deliberate semantic change: an unterminated string literal auto-closes at end of input, mirroring the module's existing policy for an unmatched `(`. Fuzz comparison against the previous tokenizer (500 k+ randomized inputs) shows byte-identical token streams for well-formed input and this as the only behavioral difference.

## Tests

- Package suite: 121 pass / 0 fail (incl. the fast-check property suite).
- New boundary cases (empty string, multiple escaped quotes, escaped backslashes) and an adversarial regression test: 50 k-char worst-case input must parse well under 2 s (measures single-digit ms; linear up to 400 k chars).

### CODE REVIEW

File: packages/template-conditions/src/parse.ts [hunk 1]
```diff
@@ -28,8 +28,19 @@ type Token =
...
+// character commits to a branch, that branch cannot fail partway through and
+// backtrack into a different split of the input. The quoted-string literal
+// used to be a branch of this same regex (`"[^"\\]*(?:\\.[^"\\]*)*"`), but a
+// closing quote is not guaranteed to exist, so an unterminated string forces
+// the (unanchored) `matchAll` scan below to retry an expensive "find the
+// close, honoring escapes" search starting at every subsequent `"` in the
+// input — quadratic on adversarial input. `scanString` below handles quoted
+// strings with a dedicated, guaranteed-single-pass scan instead, so this
+// regex only ever needs to match (or fail to match, in O(1)) at the exact
+// position it is asked to start from.
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/template-conditions/src/parse.ts:41]: Reworded without the em dashes in e63bb85f3.

File: packages/template-conditions/src/parse.ts [hunk 2]
```diff
@@ -42,38 +53,79 @@ const COMPARE_SYMBOL_TO_OP: Record<string, CompareOp> = {
...
+
+/**
+ * Scan a `"..."` literal starting at `expr[start]` (the opening quote),
+ * honoring `\\`-escapes without interpreting them. A single linear pass:
+ * each character is visited at most once, so this is safe on adversarial
+ * input regardless of how many `"` or `\\` characters it contains.
+ *
+ * A closing quote is not required — consistent with the module's
+ * degrade-gracefully policy for an unmatched `(`, a literal missing its
+ * closing quote simply closes at end of input.
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/template-conditions/src/parse.ts]: Reworded without the em dashes in e63bb85f3.

File: packages/template-conditions/src/parse.test.ts [hunk 3]
```diff
@@ -200,3 +227,33 @@ describe("degenerate input", () => {
...
   });
 });
+
+describe("tokenizer performance", () => {
+  // The tokenizer's string-literal handling used to be a regex branch
+  // (`"[^"\\]*(?:\\.[^"\\]*)*"`) scanned via an unanchored `matchAll`. On an
+  // unterminated string, the scan for a closing quote fails, and `matchAll`
+  // retries that same expensive "find the close, honoring escapes" search
+  // starting at every subsequent `"` character in the input — quadratic in
+  // the input length. Guard against that shape reappearing by asserting a
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/template-conditions/src/parse.test.ts]: Reworded without the em dashes in e63bb85f3.

File: packages/template-conditions/src/parse.ts [hunk 4]
```diff
@@ -42,38 +53,79 @@ const COMPARE_SYMBOL_TO_OP: Record<string, CompareOp> = {
...
+    raw === "==" ||
+    raw === "!=" ||
+    raw === ">" ||
+    raw === "<" ||
+    raw === ">=" ||
+    raw === "<=" ||
+    raw === "contains"
+  ) {
+    return { type: "op", raw };
+  }
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/template-conditions/src/parse.ts]: Good call: switched to Object.hasOwn against COMPARE_SYMBOL_TO_OP in e63bb85f3; package tests stay 121/0.
